doc/developer/design/20230829_topk_size_hint.md

```diff
@@ -0,0 +1,426 @@
+    OPTIONS (EXPECTED GROUP SIZE = 1000)
+    ORDER BY teacher_id, id
+    LIMIT 2
+);
+
+EXPLAIN PLAN FOR MATERIALIZED VIEW nested_distinct_on_group_by_limit;
+                                      Optimized Plan
+------------------------------------------------------------------------------------------
+ materialize.public.nested_distinct_on_group_by_limit:                                   +
+   Return                                                                                +
+     Union                                                                               +
+       Get l0                                                                            +
+       Map (null, null, null)                                                            +
+         Union                                                                           +
+           Negate                                                                        +
+             Project ()                                                                  +
+               Get l0                                                                    +
+           Constant                                                                      +
+             - ()                                                                        +
+   With                                                                                  +
+     cte l0 =                                                                            +
+       Reduce aggregates=[sum(#0), sum(#1), sum(#2)]                                     +
+         TopK order_by=[#1 asc nulls_last, #0 asc nulls_last] limit=2 exp_group_size=1000+
+           TopK group_by=[#1] order_by=[#0 asc nulls_last] limit=1 exp_group_size=1000   +
+             Reduce group_by=[#0, #1] aggregates=[max(#2)] exp_group_size=1000           +
+               Project (#0..=#2)                                                         +
+                 Get materialize.public.sections                                         +
+
+(1 row)
+```
+
+To disambiguate the query hints when necessary, we argue for an approach with the
+following characteristics:
+1. Maintain backwards compatibility with `EXPECTED GROUP SIZE` whenever there is no ambiguity, i.e.,
+if only the `EXPECTED GROUP SIZE` is specified, it is attached to all instances of reductions and
+top-k operators.
+2. When there is ambiguity, allow for the specification of additional query hints that attach
+to specific clauses in the query block, overriding the `EXPECTED GROUP SIZE` (if any is given).
+
+To operationalize the above, two additional disambiguation hints are proposed:
+
+1. `LIMIT GROUP SIZE`: This hint attaches to the `TopK` operator implementing the `LIMIT` clause.
+2. `DISTINCT ON GROUP SIZE`: This hint attaches to the `TopK` operator implementing the `DISTINCT ON` clause.
+
+As implied above, if no `EXPECTED GROUP SIZE` is given but one of `LIMIT GROUP SIZE` or
```

Comment:
I'm not sure if we can have a `TopK` coming from a `LIMIT` clause with a non-empty `group_key`. If this is the case, maybe a less confusing name is `LIMIT INPUT SIZE`.

Reply:
AFAIK, this can happen either by a `LATERAL` join or by a pattern like the one below:
```sql
CREATE MATERIALIZED VIEW alt_section_of_top_course_per_teacher AS
SELECT id AS teacher_id, 
       (SELECT id AS section_id 
        FROM sections
        WHERE teacher_id = teachers.id
        OPTIONS (EXPECTED GROUP SIZE = 1000)
        ORDER BY course_id DESC
        LIMIT 1)
FROM teachers;
```

Reply:
I think in this case the raw plan still contains `TopK` with an empty `group_key`, and the key is populated when we decorrelate and move from HIR to MIR.

Reply:
That is true, but when folks look at the optimized plan, there'll be an `exp_group_size` listed there. So one might wonder why there is a group if the hint name did not have one. No strong feelings from my side, so if you feel this would improve understandability, I am happy to change the name to `LIMIT INPUT SIZE` and drop the `GROUP` qualifier in this case.

Reply:
> No strong feelings from my side

Same here.